### PR TITLE
Damage/Health and BodyPart Code

### PR DESCRIPTION
**Pull request by L00TG0BLIN**
Current Respective Functionality:
Health/Damage/Threshold: Updates information on limbs, their health based on damage types, and the states of each limb (i.e. burned, bruised, bleeding, etc)
BodyPart/BodyPartBase: Handles the severance of limbs as well as establishes the in-game reciprocal to the logical limbs
Current Severing Functionality: Can instantiate single body part based on specification 
TO DO: 
Update Health/Damage/Threshold so that it corresponds to Mirror Code such that it can properly utilize BodyPart code (I think that's how Mirror works, correct me if I'm wrong)
Edit BodyPartBase Code so that severing a limb also severs the extension if one exists (i.e. hand, foot) and makes sure they are properly attached
Edit BodyPartBase Code so that the limb being severed turns invisible when the instantiation occurs
TO IGNORE:
There is Blood Code in Health that should be ignored for the time being, some possible changes have been mentioned that have not been made yet

Summary: There is still a bit of work to do on BodyPart Base, but enough work has been done to create the weapon stats as well as some other health-related functions (i.e. basic medical)

### CODE REVIEW

File: Assets/Scripts/Player/Skeleton/Health.cs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+                //Calculates if a limb is crippled
+                if (target != (int)SkeletalSections.torso && !statusBools[target].detached && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minDetached, Thresholds.thresholds[(int)Entities.Organic].maxDetached) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].brute)
+                {
+                    statusBools[target].detached = true;
+                    //BodyPart.severMe();
+                }
+            }
+        }
+    }
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Health.cs]: Function name should be in PascalCase.

Additionally, it is incredibly hard to follow what is going on here. Nested if statements should be avoided like the plague. Long conditions should be split up and/or moved into their own functions. And things should be named more explicitly! Additionally, the code for crippling checks is repeated. Here is how I would rewrite this section:

```
        private void BruteDamageEvaluation(int target, Damage.weaponTags[] tags)
        {
            for (int i = 0; i < tags.Length; i++)
            {
                switch (tags[i])
                {
                    case (int) Damage.weaponTags.blunt:
                        EvaluateBruise(target);
                        EvaluateCripple(target);
                        break;
                    case (int) Damage.weaponTags.stab:
                    case (int) Damage.weaponTags.cut:
                        EvaluateBleed(target);
                        EvaluateCripple(target);
                        break;
                    default:
                        Debug.LogError($"Unknown weapon tag {tags[i]} when evaluating brute damage status.");
                        break;
                }
            }
        }

        private void EvaluateBruise(int target)
        {
            //Already bruised, ignore.
            if (statusBools[target].bruised)
            {
                return;
            }

            float bruiseAmount = Random.Range(Thresholds.thresholds[(int) Entities.Organic].minBruise,
                Thresholds.thresholds[(int) Entities.Organic].maxBruise);

            //Not enough damage to bruise, ignore.
            if (bruiseAmount * limbHealthInfo[target].maxLimbHP >= limbHealthInfo[target].brute)
            {
                return;
            }

            statusBools[target].bruised = true;
        }

        private void EvaluateCripple(int target)
        {
            //Already crippled, ignore.
            if (statusBools[target].crippled)
            {
                return;
            }

            //Detatched limb can't be crippled, ignore.
            if (statusBools[target].detached)
            {
                return;
            }

            float crippleAmount = Random.Range(Thresholds.thresholds[(int) Entities.Organic].minCrippled,
                Thresholds.thresholds[(int) Entities.Organic].maxCrippled);

            //Not enough damage to cripple, ignore.
            if (crippleAmount * limbHealthInfo[target].maxLimbHP <= limbHealthInfo[target].brute)
            {
                return;
            }
            
            statusBools[target].crippled = true;
        }

        private void EvaluateBleed(int target)
        {
            //Already bleeding, ignore.
            if (statusBools[target].bleeding)
            {
                return;
            }

            float bleedAmount = Random.Range(Thresholds.thresholds[(int) Entities.Organic].minBleed,
                Thresholds.thresholds[(int) Entities.Organic].maxBleed);
            
            //Not enough damage to bleed, ignore.
            if (bleedAmount * limbHealthInfo[target].maxLimbHP <= limbHealthInfo[target].brute)
            {
                return;
            }
            
            statusBools[target].bleeding = true;
            numBleedingLimbs++;
            /*if(numBleedingLimbs == 1)
            StartCoroutine("BloodLoss");
            StartCoroutine("BloodRegen");
            */
        }
```

This can be further improved. The `damageAmount` variables and their associated calculations are nearly identical in all 3 functions, they could be generalized. But that's not important as it is likely to change in the future.

File: Assets/Scripts/Player/Skeleton/Health.cs [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+    private void burnDmgEval(int target, Damage.weaponTags[] tags)
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Health.cs]: Should be rewritten in the same way as given in the above example.

File: Assets/Scripts/Player/Skeleton/Health.cs [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        }
+
+        //Calculates if a limb is blistered
+        if (!statusBools[target].blistered && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minBlistered, Thresholds.thresholds[(int)Entities.Organic].maxBlistered) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].burn)
+        {
+            statusBools[target].blistered = true;
+        }
+    }
+
+    private void critEval()
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Health.cs]: PascalCase

File: Assets/Scripts/Player/Skeleton/Health.cs [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+    //Coroutine to regain blood overtime. Currently 2 cc's of blood are gained every 7 seconds
+    private IEnumerator BloodRegen()
+    {
+        while(currentBloodVol < maxBloodVol - 1)
+        {
+            currentBloodVol += 2;
+            yield return new WaitForSecondsRealtime(7);
+        }
+        if (currentBloodVol == maxBloodVol - 1)
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Health.cs]: missing { } brackets

File: Assets/Scripts/Player/Skeleton/Health.cs [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+            currentBloodVol++;
+    }
+
+    private void dmgDebug()
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Health.cs]: PascalCase

File: Assets/Scripts/Player/Skeleton/Health.cs [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    {
+
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Health.cs]: Remove whitespace at start and end of function

File: Assets/Scripts/Player/Skeleton/Hitbox.cs.obsolete [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+			
+		}
+
+		// Update is called once per frame
+		void Update()
+		{
+			
+		}
+	}
+}
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Hitbox.cs.obsolete]: Why commit unused scripts?

File: Assets/Scripts/Player/Skeleton/Hurtbox.cs.obsolete [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+				if (hitbox != null & ableToHit) {
+					//Wham!
+					hitbox.bp.damageMe("Brute", damageAmount);
+					ableToHit = false;
+					Invoke("SetAbleToHit", invincibleTime);
+				}
+			}
+		}
+	}
+}
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Hurtbox.cs.obsolete]: Remove unused script.

File: Assets/Scripts/Player/Skeleton/Thresholds.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        thresholds[(int)Entities.Organic].maxBurn = .25f;
+        thresholds[(int)Entities.Organic].minNumb = .35f;
+        thresholds[(int)Entities.Organic].maxNumb = .45f;
+        thresholds[(int)Entities.Organic].minBlistered = .55f;
+        thresholds[(int)Entities.Organic].maxBlistered = .75f;
+        //critStatus
+        thresholds[(int)Entities.Organic].softCrit = -0.5f;
+        thresholds[(int)Entities.Organic].hardCrit = -1f;
+    }
+}
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Skeleton/Thresholds.cs]: Identical script as the one in Player/Thresholds.cs. One of these should be deleted. Additionally, why do the values need to be initialized at runtime if they are static anyway?

Just set them immediately when declaring the `new DamageThresholds[7];`;

**Comment by Dyhr** [comment]
Superseded by #92
